File: app/analyzers/m3_auth.py
"""
M3 -- SPF / DKIM / DMARC re-verification.

Important: do NOT trust the Authentication-Results header. Below the trust
boundary it is attacker-supplied like everything else. Re-verify from
email.raw_bytes, which M1 kept on the object precisely for this.

DKIM is verified cryptographically from the raw bytes (offline-capable: a
missing or broken signature is detectable with no network). SPF and DMARC need
DNS; when it is unreachable they degrade to "not checked", never to a guess.

DMARC is a PASS/FAIL of alignment, never of policy strength: a domain that
publishes p=reject is well-secured, not suspicious. We only raise
dmarc_fail_strict when authentication actually fails alignment AND the domain
asked us to reject/quarantine such mail.

Owner: Track B
"""

from __future__ import annotations

import re
from uuid import UUID

try:
    import spf  # pyspf
except ImportError:
    spf = None

try:
    import dkim  # dkimpy -- the import name is 'dkim', NOT 'dkimpy'
except ImportError:
    dkim = None

try:
    import dns.resolver  # dnspython, for the DMARC TXT lookup
except ImportError:
    dns = None

from app.analyzers.base import register
from app.analyzers.m2_headers import authenticated_origin
from app.schemas.email import ParsedEmail
from app.schemas.evidence import Analyzer, Evidence

_DMARC_P_RE = re.compile(r"[;\s]p\s*=\s*(\w+)", re.IGNORECASE)
_DKIM_D_RE = re.compile(r"[;\s]d\s*=\s*([^;]+)", re.IGNORECASE)


def _domain_of(addr: str | None) -> str:
    if not addr or "@" not in addr:
        return ""
    return addr.rsplit("@", 1)[-1].strip().lower().rstrip(">")


def _aligned(a: str, b: str) -> bool:
    """Relaxed alignment: same domain, or one is the other's organizational parent."""
    if not a or not b:
        return False
    a, b = a.lower(), b.lower()
    return a == b or a.endswith("." + b) or b.endswith("." + a)


# --- SPF ---------------------------------------------------------------------
def _verify_spf(case_id: UUID, email: ParsedEmail, client_ip: str | None) -> tuple[Evidence | None, bool]:
    """Return (hard-fail evidence or None, spf_aligned_pass).

    client_ip is the IP that connected to our trust boundary (M2). When it is
    known we check SPF against it; otherwise SPF is skipped rather than guessed.
    """
    if spf is None:
        return None, False
    try:
        ip = client_ip
        if not ip:
            return None, False

        # Envelope sender (MAIL FROM) domain drives SPF; fall back to From.
        mail_from_domain = _domain_of(email.return_path) or _domain_of(email.from_addr)
        if not mail_from_domain:
            return None, False

        result, _ = spf.check2(i=ip, s=f"user@{mail_from_domain}", h=mail_from_domain)
        aligned_pass = result == "pass" and _aligned(mail_from_domain, _domain_of(email.from_addr))

        if result == "fail":
            return (
                Evidence.triggered(
                    case_id, Analyzer.M3_AUTH, "spf_fail_hard",
                    detail={"ip": ip, "domain": mail_from_domain,
                            "explanation": "SPF hard fail: this IP is not authorized to send "
                                           "for the envelope-sender domain."},
                ),
                aligned_pass,
            )
        return None, aligned_pass
    except Exception:
        return None, False


# --- DKIM --------------------------------------------------------------------
def _dkim_signing_domains(email: ParsedEmail) -> list[str]:
    return [
        m.group(1).strip().lower()
        for k, v in email.headers
        if k.lower() == "dkim-signature"
        for m in [_DKIM_D_RE.search(v)]
        if m
    ]


def _verify_dkim(case_id: UUID, email: ParsedEmail) -> tuple[Evidence | None, bool]:
    """Return (evidence or None, dkim_aligned_pass). No signature -> (None, False)."""
    if dkim is None or not email.raw_bytes:
        return None, False

    sig_domains = _dkim_signing_domains(email)
    if not sig_domains:
        return None, False  # unsigned: not scored (see module docstring)

    try:
        ok = bool(dkim.verify(email.raw_bytes))
    except Exception:
        ok = False

    d = sig_domains[0]
    from_domain = _domain_of(email.from_addr)
    aligned = _aligned(d, from_domain)

    if ok and aligned:
        return (
            Evidence.triggered(
                case_id, Analyzer.M3_AUTH, "dkim_valid_aligned",
                detail={"domain": d,
                        "explanation": f"DKIM signature valid and d={d} aligns with the From domain."},
            ),
            True,
        )
    if not ok:
        return (
            Evidence.triggered(
                case_id, Analyzer.M3_AUTH, "dkim_fail",
                detail={"domain": d,
                        "explanation": f"DKIM signature (d={d}) failed cryptographic verification "
                                       "— content was altered or the signature was forged."},
            ),
            False,
        )
    # valid but not aligned (mailing list / ESP): no credit, no penalty.
    return None, False


# --- DMARC -------------------------------------------------------------------
def _dmarc_policy(domain: str) -> str | None:
    """Return the DMARC policy word (none/quarantine/reject) via a DNS tree walk."""
    if dns is None or not domain:
        return None
    labels = domain.split(".")
    # Walk up: exact domain, then progressively broader parents (RFC 9989 tree walk).
    for i in range(len(labels) - 1):
        candidate = ".".join(labels[i:])
        try:
            answers = dns.resolver.resolve(f"_dmarc.{candidate}", "TXT", lifetime=5)
        except Exception:
            continue
        for rr in answers:
            txt = b"".join(getattr(rr, "strings", [])).decode(errors="replace") or str(rr).strip('"')
            if txt.lower().startswith("v=dmarc1"):
                m = _DMARC_P_RE.search(txt)
                return (m.group(1).lower() if m else "none")
    return None


def _verify_dmarc(case_id: UUID, email: ParsedEmail, authenticated: bool) -> Evidence | None:
    from_domain = _domain_of(email.from_addr)
    if not from_domain:
        return None
    policy = _dmarc_policy(from_domain)
    if policy is None:
        return None  # no policy reachable/published: not a failure by itself
    # A strict policy is only a FINDING when authentication actually failed.
    if policy in ("reject", "quarantine") and not authenticated:
        return Evidence.triggered(
            case_id, Analyzer.M3_AUTH, "dmarc_fail_strict",
            detail={"domain": from_domain, "policy": policy,
                    "explanation": f"Neither SPF nor DKIM aligned with the From domain, and the "
                                   f"domain's DMARC policy is p={policy} — this message violates "
                                   "a policy the domain owner explicitly published."},
        )
    return None


@register(Analyzer.M3_AUTH)
async def analyze(case_id: UUID, email: ParsedEmail) -> list[Evidence]:
    """Re-verify SPF, DKIM, and DMARC from the raw bytes and DNS."""
    if spf is None and dkim is None:
        return [
            Evidence.unavailable(
                case_id, Analyzer.M3_AUTH, "dkim_valid_aligned",
                "M3 unavailable: neither pyspf nor dkimpy is installed.",
            )
        ]

    ev: list[Evidence] = []

    # The IP that connected to our trust boundary is the correct SPF client IP.
    boundary, _hops = authenticated_origin(email)
    client_ip = boundary.from_ip if boundary is not None else None

    spf_ev, spf_pass = _verify_spf(case_id, email, client_ip)
    if spf_ev:
        ev.append(spf_ev)

    dkim_ev, dkim_pass = _verify_dkim(case_id, email)
    if dkim_ev:
        ev.append(dkim_ev)

    # DMARC passes if EITHER SPF or DKIM produced an aligned pass.
    dmarc_ev = _verify_dmarc(case_id, email, authenticated=spf_pass or dkim_pass)
    if dmarc_ev:
        ev.append(dmarc_ev)

    # Only negative/credit signals above are scored. If none fired, we still
    # verified what we could -- record a CLEAR so the log shows the check ran.
    triggered = [e for e in ev if e.signal != "dkim_valid_aligned"]
    if not triggered:
        ev.append(
            Evidence.clear(
                case_id, Analyzer.M3_AUTH, "auth_verification_passed",
                detail={"explanation": "No SPF/DKIM/DMARC alignment failures detected "
                                       "(re-verified from raw bytes; DNS-dependent checks "
                                       "skipped when unreachable)."},
            )
        )

    return ev

File: web/src/lib/analyzers/m3_auth.ts
/**
 * M3 -- SPF / DKIM / DMARC re-verification (ported from app/analyzers/m3_auth.py).
 *
 * Do NOT trust the Authentication-Results header -- below the trust boundary it
 * is attacker-supplied. Re-verify from rawBytes (kept by M1). DKIM is verified
 * cryptographically (offline-capable); SPF and DMARC need DNS and degrade to
 * "not checked" when unreachable, never to a guess.
 *
 * DMARC is a PASS/FAIL of alignment, never of policy strength: dmarc_fail_strict
 * fires only when auth actually fails alignment AND the domain published p=reject
 * or p=quarantine.
 */

import { resolveTxt } from 'node:dns/promises';
import { dkimVerify } from 'mailauth/lib/dkim/verify';
import { spf as mailauthSpf } from 'mailauth/lib/spf';
import { arc as arcVerify } from 'mailauth/lib/arc';
import { Analyzer, Status, clear, triggered, unavailable, type Evidence } from '../schemas/evidence';
import { registrableDomain, sameOrgDomain } from './domains';
import { headerValues, type ParsedEmail } from '../schemas/email';
import { register } from './base';
import { authenticatedOrigin } from './m2_headers';
import { DKIM_DEMO_KEYS } from './dkim_demo_keys';

// Resolver for DKIM key lookups: serve bundled demo keys locally (so the
// self-contained signed samples verify without DNS), else use real DNS.
export async function dkimResolver(name: string, rr: string): Promise<string[][]> {
  if (rr !== 'TXT') return [];
  // normalize: some runtimes hand the resolver an FQDN with a trailing dot / mixed case
  const key = name.replace(/\.$/, '').toLowerCase();
  if (DKIM_DEMO_KEYS[key]) return [[DKIM_DEMO_KEYS[key]]];
  try {
    return await resolveTxt(key);
  } catch {
    return [];
  }
}

const DMARC_P_RE = /[;\s]p\s*=\s*(\w+)/i;
const DMARC_SP_RE = /[;\s]sp\s*=\s*(\w+)/i;
const DKIM_D_RE = /[;\s]d\s*=\s*([^;]+)/i;

export type TxtResolver = (name: string) => Promise<string[][]>;

function domainOf(addr: string | null): string {
  if (!addr || !addr.includes('@')) return '';
  return addr.split('@').pop()!.trim().toLowerCase().replace(/>$/, '');
}

/** DMARC relaxed alignment: identical, or the same registrable (organisational)
 *  domain per the Public Suffix List. Uses PSL private suffixes, so a signature
 *  from a shared platform (d=herokuapp.com) does NOT align with a tenant
 *  (From @victim.herokuapp.com) -- the naive suffix match used to. */
export function aligned(a: string, b: string): boolean {
  return sameOrgDomain(a, b);
}

export function dkimSigningDomains(email: ParsedEmail): string[] {
  const out: string[] = [];
  for (const v of headerValues(email, 'dkim-signature')) {
    const m = DKIM_D_RE.exec(v);
    if (m) out.push(m[1].trim().toLowerCase());
  }
  return out;
}

// --- DKIM --------------------------------------------------------------------
export async function verifyDkim(
  caseId: string,
  email: ParsedEmail,
  resolver?: (name: string, rr: string) => Promise<string[][]>,
): Promise<{ ev: Evidence | null; pass: boolean }> {
  if (!email.rawBytes || !email.rawBytes.length) return { ev: null, pass: false };
  const sigDomains = dkimSigningDomains(email);
  if (!sigDomains.length) return { ev: null, pass: false }; // unsigned: not scored

  const fromDomain = domainOf(email.from_addr);
  let results: Array<{ signingDomain?: string; status?: { result?: string } }> = [];
  let threw = false;
  try {
    const res = await dkimVerify(email.rawBytes, resolver ? { resolver } : {});
    results = (res?.results ?? []) as Array<{ signingDomain?: string; status?: { result?: string } }>;
  } catch {
    threw = true;
  }

  // Only a signature whose d= ALIGNS with the From domain can authenticate this
  // message. A valid signature from an unrelated domain (attacker.com) must NOT
  // credit a From: victim.com message -- that was the cross-domain forgery gap.
  const domOf = (r: { signingDomain?: string }) => (r.signingDomain || '').toLowerCase();
  const alignedResults = results.filter((r) => aligned(domOf(r), fromDomain));
  const alignedHas = (result: string) => alignedResults.some((r) => r.status?.result === result);
  const alignedDom =
    alignedResults.find((r) => domOf(r))?.signingDomain
    ?? sigDomains.find((d) => aligned(d, fromDomain))
    ?? sigDomains[0];
  const alignedSigExists = sigDomains.some((d) => aligned(d, fromDomain));

  if (alignedHas('pass')) {
    return {
      ev: triggered(caseId, Analyzer.M3_AUTH, 'dkim_valid_aligned', {
        domain: alignedDom,
        explanation: `DKIM signature valid and d=${alignedDom} aligns with the From domain.`,
      }),
      pass: true,
    };
  }
  if (alignedSigExists) {
    // An aligned signature that could NOT be verified because the key/DNS lookup
    // failed (offline, temperror, or the whole verify threw, or mailauth produced
    // no result for it) is evidence-UNAVAILABLE -- never a silent fail or pass.
    if (threw || alignedHas('temperror') || alignedResults.length === 0) {
      return {
        ev: unavailable(caseId, Analyzer.M3_AUTH, 'dkim_unavailable',
          `DKIM for d=${alignedDom} could not be verified (offline or DNS/key lookup failed).`),
        pass: false,
      };
    }
    // Aligned signature present and evaluated, but not 'pass' (fail / neutral /
    // permerror = broken body hash or forged/altered signature): an integrity fail.
    return {
      ev: triggered(caseId, Analyzer.M3_AUTH, 'dkim_fail', {
        domain: alignedDom,
        explanation: `DKIM signature (d=${alignedDom}) failed verification — content was altered or the signature was forged.`,
      }),
      pass: false,
    };
  }
  return { ev: null, pass: false }; // only NON-aligned signatures (e.g. a valid attacker.com sig on a victim.com From): no credit, no penalty
}

// --- SPF ---------------------------------------------------------------------
export async function verifySpf(
  caseId: string,
  email: ParsedEmail,
  clientIp: string | null,
): Promise<{ ev: Evidence | null; pass: boolean }> {
  if (!clientIp) return { ev: null, pass: false };
  const mailFromDomain = domainOf(email.return_path) || domainOf(email.from_addr);
  if (!mailFromDomain) return { ev: null, pass: false };
  try {
    const res = await mailauthSpf({
      ip: clientIp,
      helo: mailFromDomain,
      sender: `user@${mailFromDomain}`,
      mta: 'mailtrace',
    });
    const result = (res?.status?.result ?? '').toLowerCase();
    const alignedPass = result === 'pass' && aligned(mailFromDomain, domainOf(email.from_addr));
    if (result === 'fail') {
      return {
        ev: triggered(caseId, Analyzer.M3_AUTH, 'spf_fail_hard', {
          ip: clientIp,
          domain: mailFromDomain,
          explanation: 'SPF hard fail: this IP is not authorized to send for the envelope-sender domain.',
        }),
        pass: alignedPass,
      };
    }
    if (result === 'temperror' || result === 'permerror') {
      return {
        ev: unavailable(caseId, Analyzer.M3_AUTH, 'spf_unavailable',
          `SPF for ${mailFromDomain} could not be evaluated (${result} — DNS unreachable).`),
        pass: false,
      };
    }
    return { ev: null, pass: alignedPass };
  } catch (e) {
    if (isDnsNetworkError(e)) {
      return {
        ev: unavailable(caseId, Analyzer.M3_AUTH, 'spf_unavailable', 'SPF could not be evaluated (DNS unreachable).'),
        pass: false,
      };
    }
    return { ev: null, pass: false };
  }
}

// --- ARC ---------------------------------------------------------------------
// Verify the Authenticated Received Chain. cv=pass means each ARC hop validated
// the previous one, so the authentication results recorded before forwarding are
// trustworthy -- the legitimate reason a forwarded message fails SPF at the final
// hop. 'unavailable' = the ARC keys/DNS could not be reached (offline), so we
// must NOT use it to excuse an SPF failure.
export type ArcStatus = 'pass' | 'fail' | 'none' | 'unavailable';

export async function verifyArc(email: ParsedEmail, resolver?: (name: string, rr: string) => Promise<string[][]>): Promise<ArcStatus> {
  if (!email.rawBytes?.length) return 'none';
  try {
    const dk = await dkimVerify(email.rawBytes, resolver ? { resolver } : {});
    const chain = (dk as { arc?: { chain?: unknown[] } })?.arc;
    if (!chain || !Array.isArray(chain.chain) || chain.chain.length === 0) return 'none';
    const res = await arcVerify(chain as Parameters<typeof arcVerify>[0], resolver ? { resolver } : {});
    const r = (res as { status?: { result?: string } })?.status?.result;
    return r === 'pass' || r === 'fail' || r === 'none' ? r : 'none';
  } catch {
    return 'unavailable'; // ARC key lookup / DNS failed -- cannot confirm, so do not excuse SPF
  }
}

// --- DMARC -------------------------------------------------------------------
// A DNS resolver error that means "couldn't reach DNS" (offline / transient),
// as opposed to a definitive "no such record" (ENOTFOUND / ENODATA). The former
// must surface as UNAVAILABLE so a policy violation is never silently masked.
const DNS_NET_ERRORS = new Set(['EAI_AGAIN', 'ETIMEOUT', 'ETIMEDOUT', 'ESERVFAIL', 'ECONNREFUSED', 'ECONNRESET', 'ENETUNREACH', 'ENOTIMP']);
function isDnsNetworkError(e: unknown): boolean {
  const code = (e as { code?: string })?.code;
  return code ? DNS_NET_ERRORS.has(code) : false;
}

export interface DmarcLookup { policy: string | null; unavailable: boolean }

async function lookupDmarc(domain: string, txt: TxtResolver): Promise<{ found: boolean; p: string | null; sp: string | null; netErr: boolean }> {
  try {
    const records = await txt(`_dmarc.${domain}`);
    for (const rr of records) {
      const s = rr.join('');
      if (s.toLowerCase().startsWith('v=dmarc1')) {
        const pm = DMARC_P_RE.exec(s);
        const spm = DMARC_SP_RE.exec(s);
        return { found: true, p: pm ? pm[1].toLowerCase() : 'none', sp: spm ? spm[1].toLowerCase() : null, netErr: false };
      }
    }
    return { found: false, p: null, sp: null, netErr: false };
  } catch (e) {
    return { found: false, p: null, sp: null, netErr: isDnsNetworkError(e) };
  }
}

// DMARC record discovery per RFC 7489 / 9989: the EXACT From domain, then the
// organisational domain. A subdomain with no record of its own is governed by
// the org domain's sp= (subdomain policy) when present, else its p=. (No walk of
// arbitrary intermediate labels, which real DMARC does not consult.)
export async function dmarcPolicy(domain: string, txt: TxtResolver = defaultTxt): Promise<DmarcLookup> {
  if (!domain) return { policy: null, unavailable: false };
  const exact = await lookupDmarc(domain, txt);
  if (exact.found) return { policy: exact.p, unavailable: false };
  let netErr = exact.netErr;
  const org = registrableDomain(domain);
  if (org && org !== domain) {
    const orgRec = await lookupDmarc(org, txt);
    if (orgRec.found) return { policy: orgRec.sp ?? orgRec.p, unavailable: false };
    netErr = netErr || orgRec.netErr;
  }
  // No policy found. If DNS was actually unreachable we cannot conclude "no
  // policy"; report unavailable so the caller emits UNAVAILABLE, not a clean pass.
  return { policy: null, unavailable: netErr };
}

async function defaultTxt(name: string): Promise<string[][]> {
  return resolveTxt(name);
}

export async function verifyDmarc(
  caseId: string,
  email: ParsedEmail,
  authenticated: boolean,
  txt: TxtResolver = defaultTxt,
): Promise<Evidence | null> {
  const fromDomain = domainOf(email.from_addr);
  if (!fromDomain) return null;
  const { policy, unavailable: dmarcDown } = await dmarcPolicy(fromDomain, txt);
  if (dmarcDown) {
    return unavailable(caseId, Analyzer.M3_AUTH, 'dmarc_unavailable',
      `DMARC policy for ${fromDomain} could not be retrieved (DNS unreachable) — a policy violation cannot be ruled out.`);
  }
  if (policy === null) return null; // no policy published: not a failure by itself
  if ((policy === 'reject' || policy === 'quarantine') && !authenticated) {
    return triggered(caseId, Analyzer.M3_AUTH, 'dmarc_fail_strict', {
      domain: fromDomain,
      policy,
      explanation:
        `Neither SPF nor DKIM aligned with the From domain, and the domain's DMARC policy is p=${policy} — ` +
        'this message violates a policy the domain owner explicitly published.',
    });
  }
  return null;
}

export async function analyze(caseId: string, email: ParsedEmail): Promise<Evidence[]> {
  const ev: Evidence[] = [];

  const { boundary } = authenticatedOrigin(email);
  const clientIp = boundary?.from_ip ?? null;

  let spfPass = false;
  let dkimPass = false;

  try {
    const spfR = await verifySpf(caseId, email, clientIp);
    spfPass = spfR.pass;
    if (spfR.ev?.signal === 'spf_fail_hard' && spfR.ev.status === Status.TRIGGERED) {
      // A hard SPF fail is the expected symptom of legitimate forwarding. Only
      // penalise it if a valid ARC chain does NOT vouch for the pre-forward auth.
      const arc = await verifyArc(email, dkimResolver);
      if (arc === 'pass') {
        ev.push(
          triggered(caseId, Analyzer.M3_AUTH, 'arc_authenticated', {
            explanation:
              'SPF failed at the receiving hop, but a valid ARC chain (cv=pass) shows the message was ' +
              'authenticated before it was forwarded — legitimate forwarding, not spoofing. SPF penalty suppressed.',
          }),
        );
      } else {
        ev.push(spfR.ev); // no valid ARC to excuse it — keep the SPF-fail penalty
      }
    } else if (spfR.ev) {
      ev.push(spfR.ev);
    }
  } catch {
    /* SPF is best-effort; ignore */
  }

  const dkimR = await verifyDkim(caseId, email, dkimResolver);
  if (dkimR.ev) ev.push(dkimR.ev);
  dkimPass = dkimR.pass;

  try {
    const dmarcEv = await verifyDmarc(caseId, email, spfPass || dkimPass);
    if (dmarcEv) ev.push(dmarcEv);
  } catch {
    /* DMARC DNS best-effort */
  }

  const triggeredFindings = ev.filter((e) => e.signal !== 'dkim_valid_aligned');
  if (!triggeredFindings.length) {
    ev.push(
      clear(caseId, Analyzer.M3_AUTH, 'auth_verification_passed', {
        explanation:
          'No SPF/DKIM/DMARC alignment failures detected (re-verified from raw bytes; DNS-dependent checks skipped when unreachable).',
      }),
    );
  }

  return ev;
}

register(Analyzer.M3_AUTH, analyze);
